src/Tools/dotnet-monitor/TaskCompletionSourceWithCancellation.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+#if UNITTEST
+namespace Microsoft.Diagnostics.Monitoring.TestCommon
+#else
+namespace Microsoft.Diagnostics.Tools.Monitor
+#endif
+{
+    public sealed class TaskCompletionSourceWithCancellation<T> :
+        TaskCompletionSource<T>
+    {
+        public TaskCompletionSourceWithCancellation(TaskCreationOptions creationOptions)
+            : base(creationOptions)
+        {
+        }
+
+        public async Task<T> WithCancellation(CancellationToken token)
```

Comment:
Wouldn't this be preferable as an extension method rather than an inherited class?

Reply:
The runtime has basically the same implementation apparently with a derived class: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Threading/Tasks/TaskCompletionSourceWithCancellation.cs

The one benefit is that they don't have extra allocations for the closure because it passes the TCS is the state parameter and makes use of the new Register overload in .NET 6.

However, I can still make it an extension method instead since we can't avoid the allocation.